### PR TITLE
Fix ranges for cryptography and jsonschema based on tls_cert. lib

### PR DESCRIPTION
**Pull request by phvalguima**
Currently, tls_certificates define a `PYDEPS = [cryptography, jsonschema]`.

Hence, we should not be pinning or setting subranges for these libs. As `rpds-py` v0.19 needs rustc>=1.76, this PR migrates from rustc APT to the rustup snap.

Also covers the security issue on `ops` framework: https://github.com/canonical/operator/security/advisories/GHSA-hcmv-jmqh-fjgm

### CODE REVIEW

**Comment by phvalguima** [comment]
@carlcsaposs-canonical let's see how this CI run will go. But if that works fine, I think we should just migrate to rustup snap instead.

File: charmcraft.yaml [hunk 1]
```diff
@@ -18,14 +18,15 @@ parts:
           echo 'ERROR: Use "tox run -e build-dev" instead of calling "charmcraft pack" directly' >&2
           exit 1
       fi
+      rustup default stable
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply charmcraft.yaml:21]: The `craftctl default` is a: "please, do whatever you have to do for the BUILD step". The `override-build` allows us to then add scripts before and/or after that step; or even not call it at all...

[the remark this answers is not in the document]
Comment (phvalguima) [reply charmcraft.yaml]: I see what you mean, it makes sense @carlcsaposs-canonical.

Applied the change, let's see the results on the CI.

[the remark this answers is not in the document]
Comment (phvalguima) [reply charmcraft.yaml]: yes, that should be possible

File: charmcraft.yaml [hunk 2]
```diff
@@ -18,14 +18,17 @@ parts:
...
+    override-build: |
+      rustup default stable
+      craftctl default
     charm-strict-dependencies: true
     charm-entrypoint: src/charm.py
     prime:
       - charm_version
       - workload_version
+    build-snaps:
+      - rustup/latest/stable
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply charmcraft.yaml]: @carlcsaposs-canonical Added that little fix: if that works, then we can merge this PR! Thanks!

File: charmcraft.yaml [hunk 3]
```diff
@@ -18,14 +18,17 @@ parts:
...
     prime:
       - charm_version
       - workload_version
+    build-snaps:
+      - rustup
     build-packages:
-      - cargo
       - libffi-dev
       - libssl-dev
       - pkg-config
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply charmcraft.yaml:34]: @carlcsaposs-canonical nw, but keep in mind `rustup` "installs rustc", I am not sure that rustc comes up with all its dependencies... Anyways, given we're into that, it is indeed worth testing!

[the remark this answers is not in the document]
Comment (dragomirp) [reply charmcraft.yaml:34]: All of those are in cryptography's [build requirements](https://cryptography.io/en/latest/installation/#debian-ubuntu). I'd guess ffi is used by other packages with Rust/C components.

File: pyproject.toml [hunk 4]
```diff
@@ -31,9 +31,9 @@ ops = "^2.14.1"
 # data_platform_libs/v0/upgrade.py
 # grafana_agent/v0/cos_agent.py requires pydantic <2
 pydantic = "^1.10.17, <2"
-# tls_certificates_interface/v1/tls_certificates.py
-cryptography = "^42.0.8"
-jsonschema = "^4.21.1"
+# tls_certificates_interface/v3/tls_certificates.py
+cryptography = "*"
+jsonschema = "*"
```
Comment (Mehdi-Bendriss) [inline pyproject.toml:36]: This is too broad, we need to pin

Comment (phvalguima) [reply pyproject.toml:36]: @Mehdi-Bendriss that was agreed between me and @carlcsaposs-canonical. That is b/c `tls_certificates` lib sets its own `PYDEPS` to the latest available versions for both. So, if we pin, we are effectively using a different version then what they are testing themselves.

Comment (phvalguima) [reply pyproject.toml:36]: btw, that is only updated if we push the `poetry.lock` file.

Comment (dragomirp) [reply pyproject.toml]: @Mehdi-Bendriss, the wildcard pins are in the `charm-libs` dependency group that is aimed to follow what is expected by `PYDEPS` in various charm libraries, with the actual version pins being in the main dependency group. Poetry would pick the more restrictive version pin, if they are not in conflict, so `^42.0.8` + `*` would have been the same as just `^42.0.8`.

The idea is that we should be able to automate collection at some point and resolve what the charm expects and what the libraries expect.

**Review by dragomirp — APPROVED** [review]
We've ported this pr in PG repos and have been using it for a while now without issue.

**Review by Mehdi-Bendriss — DISMISSED** [review]

File: pyproject.toml (hunk 4 above)
Comment (Mehdi-Bendriss) [reply pyproject.toml]: Thank you all for the context!

**Review by dragomirp — DISMISSED** [review]

**Review by reneradoi — COMMENTED** [review]
Hey @phvalguima Aren't those files needed? If they still are, they should be added again via the `files` part (like [here](https://github.com/canonical/opensearch-operator/commit/3a65aad1aa8163b086ba89fe39c07cf0ca70e4f0)).

Edit: Don't mind, misunderstanding from my side.

**Review by reneradoi — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**